In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("shmya_final_version_1.csv")

In [3]:
df.head(10)

date  cutlery  tips  order_price     uid  order_id
0  2022-01-17 06:38:43        3   100         1001  120967  73041443
1  2022-01-23 11:09:22        4   110          309  325757  60500835
2  2022-01-24 19:40:03        1    10          891  986730  43230531
3  2022-01-14 18:01:22        5    70          749  868648  46180789
4  2022-01-19 22:32:43        2    10          469  863523  59266169
5  2022-01-05 11:45:43        4    90          340  728911  51050143
6  2022-01-01 08:52:23        4     0            0  779134  15841612
7  2022-01-01 06:21:14        6     0            0  751242  83957554
8  2022-01-24 11:45:30        1    70          445  594275  61860247
9  2022-01-23 05:48:53        1    50          464  596696  80813955

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26057 entries, 0 to 26056
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         26057 non-null  object
 1   cutlery      26057 non-null  int64 
 2   tips         26057 non-null  int64 
 3   order_price  26057 non-null  int64 
 4   uid          26057 non-null  int64 
 5   order_id     26057 non-null  int64 
dtypes: int64(5), object(1)
memory usage: 1.2+ MB


In [5]:
ar = np.array(list(df['cutlery']))
ar.max()

6

In [6]:
# проверим order_price на NaN
df.dropna(subset=['order_price'], how='all')

date  cutlery  tips  order_price     uid  order_id
0      2022-01-17 06:38:43        3   100         1001  120967  73041443
1      2022-01-23 11:09:22        4   110          309  325757  60500835
2      2022-01-24 19:40:03        1    10          891  986730  43230531
3      2022-01-14 18:01:22        5    70          749  868648  46180789
4      2022-01-19 22:32:43        2    10          469  863523  59266169
...                    ...      ...   ...          ...     ...       ...
26052  2022-01-04 05:21:39        4    10          763  182587  85041983
26053  2022-01-18 14:43:12        3    60          790  187031  47894568
26054  2022-01-26 18:08:02        1    90          502  247419  75350869
26055  2022-01-18 05:55:09        2    80          418  474601  79438423
26056  2022-01-09 01:06:30        2    50          802  898956  87138202

[26057 rows x 6 columns]

In [7]:
# Размерности совпадают, значит пропусков в данных в сумме заказов нет

In [8]:
# оставим те заказы, где сумма заказа не равна нулю
df_2 = df[df['order_price'] != 0]
df_2

date  cutlery  tips  order_price     uid  order_id
0      2022-01-17 06:38:43        3   100         1001  120967  73041443
1      2022-01-23 11:09:22        4   110          309  325757  60500835
2      2022-01-24 19:40:03        1    10          891  986730  43230531
3      2022-01-14 18:01:22        5    70          749  868648  46180789
4      2022-01-19 22:32:43        2    10          469  863523  59266169
...                    ...      ...   ...          ...     ...       ...
26052  2022-01-04 05:21:39        4    10          763  182587  85041983
26053  2022-01-18 14:43:12        3    60          790  187031  47894568
26054  2022-01-26 18:08:02        1    90          502  247419  75350869
26055  2022-01-18 05:55:09        2    80          418  474601  79438423
26056  2022-01-09 01:06:30        2    50          802  898956  87138202

[23303 rows x 6 columns]

In [9]:
# добавим в таблицу столбец с процентным соотношением количества чаевых к сумме заказа
df_2['pct'] = (np.array(df_2['tips']) / np.array(df_2['order_price'])) * 100
df_2

<ipython-input-9-dc5b07be3095>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_2['pct'] = (np.array(df_2['tips']) / np.array(df_2['order_price'])) * 100


date  cutlery  tips  order_price     uid  order_id  \
0      2022-01-17 06:38:43        3   100         1001  120967  73041443   
1      2022-01-23 11:09:22        4   110          309  325757  60500835   
2      2022-01-24 19:40:03        1    10          891  986730  43230531   
3      2022-01-14 18:01:22        5    70          749  868648  46180789   
4      2022-01-19 22:32:43        2    10          469  863523  59266169   
...                    ...      ...   ...          ...     ...       ...   
26052  2022-01-04 05:21:39        4    10          763  182587  85041983   
26053  2022-01-18 14:43:12        3    60          790  187031  47894568   
26054  2022-01-26 18:08:02        1    90          502  247419  75350869   
26055  2022-01-18 05:55:09        2    80          418  474601  79438423   
26056  2022-01-09 01:06:30        2    50          802  898956  87138202   

             pct  
0       9.990010  
1      35.598706  
2       1.122334  
3       9.345794  
4       2.132196  
...          ...  
26052   1.310616  
26053   7.594937  
26054  17.928287  
26055  19.138756  
26056   6.234414  

[23303 rows x 7 columns]

## Простая проверка. Возьмем среднее значение чаевых в процентах тех, у кого было меньше трех комплектов приборов и больше двух и сравним

In [10]:
# у кого меньше трех комплектов
cultery1 = np.array(df_2[df_2['cutlery'] < 3]['pct'])
cultery1.mean()

8.833760753392083

In [11]:
# у кого больше двух комплектов
cultery2 = np.array(df_2[df_2['cutlery'] > 2]['pct'])
cultery2.mean()

12.338199448567519

#### Среднее значение чаевых у тех, кто брал больше двух комплектов: 12.338199448567519; у кого меньше трех: 8.833760753392083, значит гипотеза верна

## Еще проверка

### Посмотрим на 30 процентов самых больших чевых и среди них сравним количество человек с набором приборов меньше трех и больше двух

In [12]:
quant = df_2['pct'].quantile(0.7)

In [13]:
num1 = df_2[df_2['pct'] > quant]
num1[num1['cutlery'] < 3].shape[0]

1682

In [14]:
num1[num1['cutlery'] > 2].shape[0]

5308

#### Количество человек с тремя наборами и более среди 30 процентов самых больших чаевых: 5308, а тех, у кого два и один набор: 1682. Тогда гипотеза опять подтверждается

In [15]:
dates_to_del = ['01', '02', '03', '04', '05', '06', '07', '08', '09'] # дни января, которые не нужно учитывать

In [16]:
df_new = df_2[df_2['cutlery'] > 2]
df_new = df_new[df_new['order_price'] > 800]
df_new

date  cutlery  tips  order_price     uid  order_id  \
0      2022-01-17 06:38:43        3   100         1001  120967  73041443   
18     2022-01-26 13:48:12        3   150         1097  255228  33800999   
23     2022-01-26 13:22:38        4   140         1161  781632  56865155   
31     2022-01-02 21:27:39        4    10          926  148527  73472532   
36     2022-01-08 23:36:13        4    20          890  911427  18625955   
...                    ...      ...   ...          ...     ...       ...   
26027  2022-01-25 04:18:43        3    80          841  739019  85117758   
26033  2022-01-14 07:05:08        3    90          985  540487  87039706   
26034  2022-01-19 15:07:15        3   160         1202  117066  95916480   
26038  2022-01-23 09:11:45        4   140          892  958020  31282164   
26043  2022-01-31 18:36:03        4    20          942  908699  78733530   

             pct  
0       9.990010  
18     13.673655  
23     12.058570  
31      1.079914  
36      2.247191  
...          ...  
26027   9.512485  
26033   9.137056  
26034  13.311148  
26038  15.695067  
26043   2.123142  

[4776 rows x 7 columns]

In [17]:
df_new['date'].loc[0].split(' ')[0].split('-')[-1]

'17'

In [18]:
l = []
for i in list(df_new.index):
    if df_new['date'].loc[i].split(' ')[0].split('-')[1] == '01' and df_new['date'].loc[i].split(' ')[0].split('-')[-1] in dates_to_del:
        l.append(i)
len(l)

1389

In [19]:
df_all = df_new.loc[set(df_new.index) - set(l)]
df_all

date  cutlery  tips  order_price     uid  order_id  \
0      2022-01-17 06:38:43        3   100         1001  120967  73041443   
24576  2022-01-12 08:44:52        6   160         1134  188501  65842738   
16390  2022-01-16 06:19:45        3    40          872  722323  42133956   
8202   2022-01-12 08:00:06        4    70         1010  823484  39814909   
8203   2022-01-13 18:10:47        4   130         1029  988154  82320631   
...                    ...      ...   ...          ...     ...       ...   
16373  2022-01-28 20:58:16        5   110          955  711181  27799483   
24565  2022-01-10 02:14:05        4   130          998  829513  72347950   
8184   2022-01-17 04:09:45        4   160         1857  808274  16956189   
16382  2022-01-19 08:40:42        3    10          904  700804  27157720   
16383  2022-01-18 13:34:47        3   140          899  464558  72189537   

             pct  
0       9.990010  
24576  14.109347  
16390   4.587156  
8202    6.930693  
8203   12.633625  
...          ...  
16373  11.518325  
24565  13.026052  
8184    8.616047  
16382   1.106195  
16383  15.572859  

[3387 rows x 7 columns]

In [20]:
df_all.shape[0] # окончательная таблица, всего 3387 пользователей

3387